### PR TITLE
nixos/nextcloud: add extraOptions and secretFile options

### LINKED ISSUE #116189: nextcloud: how to configure caching / no extraOptions-like hook?
Currently, the nextcloud modules has a set list of config options (in `services.nextcloud.config`), and no `extraOptions`-style hook for freeform text which should be appended to nextcloud's configuration file.

In particular, I am unable to figure out how to set up redis caching with the nextcloud module — the description for [services.nextcloud.caching.redis](https://search.nixos.org/options?channel=20.09&show=services.nextcloud.caching.redis&from=0&size=50&sort=relevance&query=services.nextcloud.caching.redis) states that "you still need to enable Redis in your config.php", but there is no description of how to do that, and no corresponding option in the module.

I think the module should offer something like `services.nextcloud.config.extraOptions`, particular since this is probably also an issue for other possible configurations of nextcloud (given that its [full list of configuration paramaters](https://docs.nextcloud.com/server/14/admin_manual/configuration_server/config_sample_php_parameters.html) is prohibitively long) — or am I just missing something obvious?

### PR DESCRIPTION
**Pull request by stuebinm**
###### Motivation for this change

Issue: Fixes #116189

This adds two options to the nextcloud module, with the aim of allowing the entirety of `config.php` to be set declaratively (while allowing the user to choose for which options this makes sense, since nextcloud also keeps state in there):
1. `services.nextcloud.extraOptions`, which takes free-form options given as an attribute set, and reads them in via json from the php side (which lets us prevent syntax errors in php, if not key errors – given the full length of nextcloud's potential options, I don't think specifying them all via Nix is a viable option)
2. `services.nextcloud.secretFile`, which takes a path to a json file specifying options which a user may want to keep secret and not add to their nix store; this file is read in the same way on startup by php, and may even overwrite options set in some other way.

Additionally, adds a new test for nextcloud based on the current with-postgresql-and-redis test, but with a more declarative config instead of using `nextcloud-occ`.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/services/web-apps/nextcloud.nix
Comment (lheckemann) [inline nixos/modules/services/web-apps/nextcloud.nix:523]: These should probably be read from a file (though writing it to the store is fine here) as well, as the JSON may contain single-quote characters.

**Review by lheckemann — COMMENTED** [review]
:+1: thanks! I'm always happy to see more modules supporting keeping secrets out of the store :)

For the `extraOptions` part, I think an [RFC42](https://github.com/NixOS/rfcs/blob/master/rfcs/0042-config-option.md)-style approach would be nicer, since it integrates better with the module system, though I'd defer to people who are more involved with this module on that topic. cc @ma27 :)

File: nixos/modules/services/web-apps/nextcloud.nix
Comment (Ma27) [inline nixos/modules/services/web-apps/nextcloud.nix:525]: Is `array_push` the thing you actually need here?

```
>>> $foo=[1,2,3,4];
=> [
     1,
     2,
     3,
     4,
   ]
>>> $b=[4,5,6,7];
=> [
     4,
     5,
     6,
     7,
   ]
>>> array_push($foo,$b);
=> 5
>>> $foo
=> [
     1,
     2,
     3,
     4,
     [
       4,
       5,
       6,
       7,
     ],
   ]

```

File: nixos/modules/services/web-apps/nextcloud.nix
Comment (Ma27) [inline nixos/modules/services/web-apps/nextcloud.nix:377]: IIRC using `literalExample` makes better formatting in the manual.

File: nixos/modules/services/web-apps/nextcloud.nix
Comment (stuebinm) [reply nixos/modules/services/web-apps/nextcloud.nix]: I've replaced this with the type of pkgs.formats.json, so it should be fine now.

File: nixos/modules/services/web-apps/nextcloud.nix
Comment (stuebinm) [reply nixos/modules/services/web-apps/nextcloud.nix]: implemented that it https://github.com/NixOS/nixpkgs/pull/118093/commits/83a669a0bef463b3202a544b33efc9109605ce83

Comment (stuebinm) [reply nixos/modules/services/web-apps/nextcloud.nix]: Could you specify what you mean with RFC42-style approach? As far as I can see it suggests doing these extra-options like hooks as structured attribute sets instead of strings, which is what I have done (via json formatting, since I couldn't find any better way to serialise to php).

Apart from that, the rfc talks a lot about balancing these options with "normal" nixos module options, but I didn't really want to replace any of the other parts, since the whole module is quiet complex and I don't actually understand any php (I just needed a way for secrets & extra options in my own deployment). I guess we could rename `extraOptions` to `extraSettings` though.

File: nixos/modules/services/web-apps/nextcloud.nix
Comment (Ma27) [reply nixos/modules/services/web-apps/nextcloud.nix:379]: It's correct though becaues it gives you better formatting.

File: nixos/modules/services/web-apps/nextcloud.nix
Comment (Ma27) [inline nixos/modules/services/web-apps/nextcloud.nix:396]: Tbh I don't really like that interface. Can you at least make sure that one can write
```
{
  "foo": "bar"
}
```

Also, I'm wondering if we want to do some kind of `file_get_contents`-logic for a single option specifically. And I don't really understand why `timeout` has to be here.

File: nixos/modules/services/web-apps/nextcloud.nix
Comment (Ma27) [inline nixos/modules/services/web-apps/nextcloud.nix:490]: This should be deduplicated IMHO. A `nix_read_file($filename)` could be made for instance.

File: nixos/modules/services/web-apps/nextcloud.nix
Comment (Ma27) [inline nixos/modules/services/web-apps/nextcloud.nix:526]: This creates a store-path for no reason. Any reason why we don't use `builtins.toJSON` here? Also, `json_decode` will silently fail if something goes wrong with the decode. You may want to check with `json_last_error() !== 0` and throw an error with `json_last_error_msg()` in the message.

File: nixos/modules/services/web-apps/nextcloud.nix
Comment (Ma27) [inline nixos/modules/services/web-apps/nextcloud.nix:529]: `array_replace_recursive()` may be better here though.

File: nixos/tests/nextcloud/with-declarative-redis.nix
Comment (Ma27) [inline nixos/tests/nextcloud/with-declarative-redis.nix:2]: I like that :D

File: nixos/tests/nextcloud/with-declarative-redis.nix
Comment (Ma27) [inline nixos/tests/nextcloud/with-declarative-redis.nix:102]: I think it'd make more sense here to actually test if the redis-cache itself is working.

**Review by Ma27 — COMMENTED** [review]
Sorry, I forgot about this. Left a few more comments.

**Comment by Ma27** [comment]
@stuebinm may I ask if you're still interested in working on this? :)

File: nixos/modules/services/web-apps/nextcloud.nix
Comment (stuebinm) [reply nixos/modules/services/web-apps/nextcloud.nix]: this should already work.

The "timeout"=2 bit was a mistake (it don't think it should hurt to set it here, though using a "=" for that is definitely wrong since it's not valid json); I've removed it.

File: nixos/modules/services/web-apps/nextcloud.nix
Comment (stuebinm) [reply nixos/modules/services/web-apps/nextcloud.nix]: done (with an additional parameter $error to model error messages in case the file could not be read)

File: nixos/modules/services/web-apps/nextcloud.nix
Comment (stuebinm) [reply nixos/modules/services/web-apps/nextcloud.nix:514]: I recall the reason for this was that I couldn't figure out how to escape the json string produced by `builtins.toJSON` in a way that wouldn't crash php, since the only way to write a string in php appears to be quoting it with " which json also uses. Presumably the string could be escaped, but since I'm unfamiliar with the subtleties of php strings it seemed safer to do it this way.

(As I think I mentioned previously, I don't actually know much about php; the original motivation for this PR is that I needed a secretFile option and hacked one in – so while this should work, the php code may not necessarily be very elegant or idiomatic)

File: nixos/tests/nextcloud/with-declarative-redis.nix
Comment (stuebinm) [reply nixos/tests/nextcloud/with-declarative-redis.nix:102]: I agree. However, I am unsure how to do that, and since I copied the entire test from `with-postgresql-and-redis.nix` and just changed the redis configuration, I though the test should be sufficient.

**Comment by stuebinm** [comment]
> @stuebinm may I ask if you're still interested in working on this? :)

Huh. I replied to you a week ago, but apparently github's mailin just quietly failed on us:

> I am, but it's also not an immediate concern for me and in any case I won't have time for it until at least the next week (and possibly not for the next three weeks, depending on how other things turn out). Though tbh I also almost forgot this MR was even still open …

File: nixos/modules/services/web-apps/nextcloud.nix
Comment (Ma27) [reply nixos/modules/services/web-apps/nextcloud.nix:379]: Please use `literalExample '' /* nix value */ ''` here.

File: nixos/modules/services/web-apps/nextcloud.nix
Comment (Ma27) [inline nixos/modules/services/web-apps/nextcloud.nix:396]: (nit for formatting) please use e.g. `<programlisting>` here for better readability (there should be enough examples in the manual already :))

File: nixos/modules/services/web-apps/nextcloud.nix
Comment (Ma27) [reply nixos/modules/services/web-apps/nextcloud.nix:514]: Ahh you're right! While I'm not happy with this, this would escalate way to fast (I checked a few ideas, but I couldn't find a solution that's worth the effort).
However, it'd still be nice to have a safety guard here regarding how to behave if the JSON cannot be decoded by PHP. This could by implemented like this (right after the declaration of `$EXTRACONFIG`):

``` php
if (json_last_error() !== JSON_ERROR_NONE) {
  throw new \RuntimeException(sprintf("Cannot decode %s, because: %s", $filename, json_last_error_msg()));
}
```

File: nixos/modules/services/web-apps/nextcloud.nix
Comment (Ma27) [inline nixos/modules/services/web-apps/nextcloud.nix:518]: I'd say that in both cases, `array_replace_recursive` is the better choice here.

File: nixos/tests/nextcloud/with-declarative-redis.nix
Comment (Ma27) [reply nixos/tests/nextcloud/with-declarative-redis.nix:102]: Perhaps try to upload a file to nextcloud with `rclone` (the `basic.nix` shows how to do that) and then check if something was written to the cache (no idea if that works, but such a test IMHO makes sense to check for potential regression in the cache config in the future).

**Review by Ma27 — COMMENTED** [review]
> Huh. I replied to you a week ago, but apparently github's mailin just quietly failed on us:

@stuebinm That's funny, the same thing happened to me, only learnt that here are updates because of yesterday's discourse notification :D

**Comment by Ma27** [comment]
@stuebinm do you intend to continue working on this? :)

**Comment by Ma27** [comment]
Still relevant.

**Comment by stuebinm** [comment]
anyone want to take a look at this?

**Comment by Ma27** [comment]
@stuebinm how much is still missing here in you opinion to get it merged? It would help if you could resolve the merge conflicts and resolve all of my review comments that you consider fixed.

I don't really have the time to start yet another side-project (i.e. implementing missing stuff), but if it's just a review away from a merge after that, I'm happy to do that ;-)

**Comment by jtrees** [comment]
Landed here after failing to find `extraOptions`. Very much interested.

**Review by kevincox — APPROVED** [review]

**Comment by kevincox** [comment]
The module doesn't have any maintainers but let's bring in the package maintainers. I like this approach, structured configuration is always better than bashing strings together. Once this is merged we can convert the other options to this format.

@schneefux @bachp @globin

@stuebinm if you could update the PR to fix conflicts in the meantime that would be great. If the maintainers don't speak up I think we can merge this.

**Comment by Ma27** [comment]
> The module doesn't have any maintainers but let's bring in the package maintainers

Uhm, what about me? %) 

I'll try to be able to take a look soonish.

**Comment by kevincox** [comment]
There is no metadata listing maintainers on the module. But that is a newer thing IIRC so I'm not surprised it may be missing.

Thanks for taking a look.

File: nixos/modules/services/web-apps/nextcloud.nix
Comment (stuebinm) [reply nixos/modules/services/web-apps/nextcloud.nix]: done (though in the nix_decode_json_file function)

File: nixos/modules/services/web-apps/nextcloud.nix
Comment (stuebinm) [reply nixos/modules/services/web-apps/nextcloud.nix]: note that there's now two functions: the nix_read_file function suggested here, and a nix_decode_json_file function which does the same but also handles json decoding & potential errors

File: nixos/modules/services/web-apps/nextcloud.nix
Comment (stuebinm) [reply nixos/modules/services/web-apps/nextcloud.nix]: resolved in a different way (see your other thread further down)

File: nixos/modules/services/web-apps/nextcloud.nix
Comment (stuebinm) [reply nixos/modules/services/web-apps/nextcloud.nix]: done

File: nixos/modules/services/web-apps/nextcloud.nix
Comment (stuebinm) [reply nixos/modules/services/web-apps/nextcloud.nix]: both done

File: nixos/modules/services/web-apps/nextcloud.nix
Comment (kevincox) [inline nixos/modules/services/web-apps/nextcloud.nix:753]: Nit: Why are these different. Can we just do the one-line version for both extraconfig and secret?

Comment (stuebinm) [reply nixos/modules/services/web-apps/nextcloud.nix:752]: Indeed we can! 

I think the reason why it's like this was mostly historical — this PR is pretty old and saw a lot of changes

File: nixos/modules/services/web-apps/nextcloud.nix
Comment (stuebinm) [reply nixos/modules/services/web-apps/nextcloud.nix:789]: maybe?

I guess we could do this:

```nix
${ let extraConfig = jsonFormat.generate "nextcloud-extraOptions.json" cfg.extraOptions;
   in ''
     $CONFIG = array_replace_recursive($CONFIG, nix_decode_json_file(
     "${extraConfig}",
     "decoding the json-encoded file generated from the extraOptions failed (${extraConfig}). This is probably a bug in nixpkgs."
   ));
 ''}
```

**Comment by stuebinm** [comment]
I've resolved the threads which seemed either outdated or resolved to me @Ma27

File: nixos/modules/services/web-apps/nextcloud.nix
Comment (stuebinm) [reply nixos/modules/services/web-apps/nextcloud.nix:789]: ah — tbh i don't like writing php (never properly learnt it, and every time i touch anything things seem to go up in flames), so i tend to do as much as possible on the nix side of things. But if you write a snippet / patch for that I'll be happy to add it in

File: nixos/modules/services/web-apps/nextcloud.nix
Comment (Ma27) [inline nixos/modules/services/web-apps/nextcloud.nix:558]: Should be `literalExpression` IIRC.

File: nixos/modules/services/web-apps/nextcloud.nix
Comment (Ma27) [inline nixos/modules/services/web-apps/nextcloud.nix:771]: Shouldn't this be `nix_read_secret`?

File: nixos/modules/services/web-apps/nextcloud.nix
Comment (Ma27) [reply nixos/modules/services/web-apps/nextcloud.nix:789]: > This doesn't have %s although it can theoretically never fail. Maybe it makes more sense to just pass in "file type" such as "extraOptions file" or "secretFile file" and the function itself can generate a message and include the path?

AFAIU this is nothing an admin can break while deploying bogus configuration, but only if someone contributes code that's wrong. So I'd argue it's good enough and other failures just shouldn't pass reviews.

**Review by Ma27 — COMMENTED** [review]
Did a review, overall LGTM, just added a few minor notes.
@kevincox would you mind testing the change before merging, not sure if I'll get to it soonish :)

**Comment by kevincox** [comment]
I can probably test but it doesn't look ready yet. We have some function name errors so IIUC this doesn't evaluate.

File: nixos/modules/services/web-apps/nextcloud.nix
Comment (stuebinm) [reply nixos/modules/services/web-apps/nextcloud.nix]: changed it

File: nixos/modules/services/web-apps/nextcloud.nix
Comment (stuebinm) [reply nixos/modules/services/web-apps/nextcloud.nix]: it should be, yes! seems to have been forgotten at some point in this PR's history …

File: nixos/tests/nextcloud/default.nix
Comment (stuebinm) [reply nixos/tests/nextcloud/default.nix]: done that too

**Comment by stuebinm** [comment]
I've looked at it again, the nextcloud tests all run through fine for me, and it's reasonably up to date with the master branch. Also I've removed a space on an empty line somewhere that made one of the checks fail.

**Comment by kevincox** [comment]
Result of `nixpkgs-review pr 118093` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>

**Comment by kevincox** [comment]
I updated my server to this and used it to configure my SMTP settings. It worked like a charm.

File: nixos/modules/services/web-apps/nextcloud.nix
Comment (kevincox) [reply nixos/modules/services/web-apps/nextcloud.nix:786]: I don't like passing in an invalid format string because it can't be used. Let's either pass in something trivial like `"this should never happen %s"` or handle this properly. Honestly i don't think we need to do anything too fancy, but I don't want to merge in "wrong" code as it will confuse future readers.

Comment (stuebinm) [reply nixos/modules/services/web-apps/nextcloud.nix]: I've changed it

**Review by kevincox — APPROVED** [review]
Looks good to me. @Ma27 good to merge?

**Comment by Ma27** [comment]
Oh sorry forgot to reply, sgtm :)

**Comment by kevincox** [comment]
I figured we can fix any last things in followups :)

**Pull request merged**
